### PR TITLE
[th/host-log-run] host: log for Host.run() and extend host.Host implementation

### PR DESCRIPTION
**Pull request by thom311**
a major rework of `Host.run()`. This is how I think it should be done.

- Most importantly, (automatically) log useful messages about the command invocation. Also have various (optional) arguments, to control the logging.

- Handle binary output in a sensible way.

For full details, see the commit message.

If you currently run debug logging, it doesn't really show much. With this patch, we can see all the commands that are invoked (and their output). Quote often, the commands that we invoke are rather interesting for understanding what happens, so this provides IMO a lot of useful debugging output for free.

### CODE REVIEW

File: host.py [hunk 1]
```diff
@@ -1,24 +1,85 @@
...
 from logger import logger
 
 
+_lock = threading.Lock()
+
+_global_counter_val = 0
+
+
+def _global_counter() -> int:
+    # For each run() call, we log a message when staring the command and when
```
Comment (wizhaoredhat) [inline host.py:27]: Typo for "staring" -> "starting"
Maybe a more descriptive name than _global_counter would be better such as "_unique_log_id"? WDYT?

File: host.py [hunk 2]
```diff
@@ -1,24 +1,85 @@
...
+        return f"cmd {self.debug_str()}"
+
+
+@dataclass(frozen=True)
+class Result(BaseResult[str]):
+    pass
+
+
+@dataclass(frozen=True)
+class BinResult(BaseResult[bytes]):
```
Comment (wizhaoredhat) [inline host.py:75]: What is an example of a binary result?

File: host.py [hunk 3]
```diff
@@ -34,29 +95,217 @@ def all_ports(self) -> list[dict[str, Any]]:
         return typing.cast(list[dict[str, Any]], r)
 
     @abstractmethod
-    def run(self, cmd: str, env: dict[str, str] = os.environ.copy()) -> Result:
+    def pretty_str(self) -> str:
         pass
 
+    @typing.overload
+    def run(
```
Comment (wizhaoredhat) [inline host.py:102]: Why is there 2 "run" functions with the same arguments but different return value types?

Comment (wizhaoredhat) [reply host.py:102]: Also there is no code for these functions

File: host.py (hunk 2 above)
Comment (thom311) [reply host.py:75]: there is one in the commit message:
```
python -c 'import host; r = host.local.run("printf \"foo:\\705\""); print(r)'
```
or really anytime, you get binary result, like 
```
$ python -c 'import host; host.local.run("echo -n hello | gzip")'
2024-06-18 19:47:05 DEBUG   [th:140307334326080]: cmd[1;localhost]: call `echo -n hello | gzip`
2024-06-18 19:47:05 ERROR   [th:140307334326080]: cmd[1;localhost]: └──> `echo -n hello | gzip`: [BINARY] [UNEXPECTED_BINARY] succcess; out=b'\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03\xcbH\xcd\xc9\xc9\x07\x00\x86\xa6\x106\x05\x00\x00\x00'
```

Since we call other processes (that we don't trust), we should not assume that we get benign UTF-8.

also, CDA has `Host.read_file()` which is implemented as `f"cat {filename}"`. That cannot read binary files.

Usually, we don't call programs that output binary data. The main point here is the logging, to catch cases where we get unexpected binary output.

File: host.py (hunk 3 above)
Comment (thom311) [reply host.py:102]: it's for the `@typing.overload`.

https://docs.python.org/3/library/typing.html#overload

This describes certain combinations of signatures. In particular, this one exists so that mypy knows that `Host.run(..., text=False)` gives a binary `BinResult` and that `Host.run(..., text=True)` gives a string `Result`.

Comment (wizhaoredhat) [reply host.py:102]: It makes the code really hard to follow since it spans multiple lines and takes up a lot space in my opinion. This aspect of python typing makes the code bloated which I think python wanted to solve... ironically now it is even more bloated than C++.

File: host.py (hunk 2 above)
Comment (wizhaoredhat) [reply host.py:75]: Ah i see. It is to catch unexpected cases.

**Comment by wizhaoredhat** [comment]
LGTM.... But as I mentioned python typing is awful.
